gtsfm/frontend/matcher/cacher/matcher_cacher.py

```diff
@@ -0,0 +1,181 @@
+        if keypoints_i1.responses is not None:
+            numpy_arrays_to_hash.append(keypoints_i1.responses[:NUM_KEYPOINTS_TO_SAMPLE_FOR_HASH].flatten())
+        if keypoints_i1.scales is not None:
+            numpy_arrays_to_hash.append(keypoints_i1.scales[:NUM_KEYPOINTS_TO_SAMPLE_FOR_HASH].flatten())
+        numpy_arrays_to_hash.append(descriptors_i1[:NUM_KEYPOINTS_TO_SAMPLE_FOR_HASH].flatten())
+        numpy_arrays_to_hash.append(keypoints_i2.coordinates[:NUM_KEYPOINTS_TO_SAMPLE_FOR_HASH].flatten())
+        if keypoints_i2.responses is not None:
+            numpy_arrays_to_hash.append(keypoints_i2.responses[:NUM_KEYPOINTS_TO_SAMPLE_FOR_HASH].flatten())
+        if keypoints_i2.scales is not None:
+            numpy_arrays_to_hash.append(keypoints_i2.scales[:NUM_KEYPOINTS_TO_SAMPLE_FOR_HASH].flatten())
+        numpy_arrays_to_hash.append(descriptors_i2[:NUM_KEYPOINTS_TO_SAMPLE_FOR_HASH].flatten())
+
+        # add the shapes as a numpy array
+        numpy_arrays_to_hash.append(np.array([im_shape_i1[0], im_shape_i1[1], im_shape_i2[0], im_shape_i2[1]]))
+
+        # hash the concatenation of all the numpy arrays
+        input_key = cache_utils.generate_hash_for_numpy_array(np.concatenate(numpy_arrays_to_hash))
+
+        return "{}_{}".format(self._matcher_obj_key, input_key)
+
+    def __load_result_from_cache(
+        self,
+        keypoints_i1: Keypoints,
+        keypoints_i2: Keypoints,
+        descriptors_i1: np.ndarray,
+        descriptors_i2: np.ndarray,
+        im_shape_i1: Tuple[int, int],
+        im_shape_i2: Tuple[int, int],
+    ) -> Optional[np.ndarray]:
+        """Load cached result, if it exists."""
+        cache_path = self.__get_cache_path(
+            cache_key=self.__generate_cache_key(
+                keypoints_i1=keypoints_i1,
+                keypoints_i2=keypoints_i2,
+                descriptors_i1=descriptors_i1,
+                descriptors_i2=descriptors_i2,
+                im_shape_i1=im_shape_i1,
+                im_shape_i2=im_shape_i2,
+            )
+        )
+
+        return io_utils.read_from_compressed_file(cache_path)
+
+    def __save_result_to_cache(
+        self,
+        keypoints_i1: Keypoints,
+        keypoints_i2: Keypoints,
+        descriptors_i1: np.ndarray,
+        descriptors_i2: np.ndarray,
+        im_shape_i1: Tuple[int, int],
+        im_shape_i2: Tuple[int, int],
+        match_indices: np.ndarray,
+    ) -> None:
+        """Save the results to the cache."""
+        cache_path = self.__get_cache_path(
+            cache_key=self.__generate_cache_key(
+                keypoints_i1=keypoints_i1,
+                keypoints_i2=keypoints_i2,
+                descriptors_i1=descriptors_i1,
+                descriptors_i2=descriptors_i2,
+                im_shape_i1=im_shape_i1,
+                im_shape_i2=im_shape_i2,
+            )
+        )
+        io_utils.write_to_compressed_file(match_indices, cache_path)
+
+    def match(
```

Comment:
why match here? i thought we were just saving what the `Matcher` computed to disk, and then reloading matches "as-is"?

Reply:
We will reload if it exists. Otherwise we will compute and persist.

So the catcher exposes the same `match` API, and internally it either loads from disk or computes it.